1-3 уроки

VladislavMikh opened the pull request.

Commit:
1 урок

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -42,7 +42,7 @@ fun quadraticRootProduct(a: Double, b: Double, c: Double): Double {
     // Решаем x^2 - 3*x + 2 = 0
     val x1x2 = quadraticRootProduct(1.0, -3.0, 2.0)
-    println("Root product: $x1x2")
+    println ("Root product: $x1x2")
```
В коммитах лучше избегать ненужных изменений, так как они усложняют объединение наборов изменений от разных разработчиков.

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +60,36 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  */
-fun angleInRadian(grad: Int, min: Int, sec: Int): Double = TODO()
-
+fun angleInRadian(grad: Int, min: Int, sec: Int): Double = grad * PI / 180 + min / 60.0 * PI / 180 + sec / 3600.0 * PI / 180
```
Функцию можно упростить с помощью элементарных арифметических преобразований.

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +60,36 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  */
-fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double = TODO()
+fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double {
+    val change_x_sqr = ( x2 - x1 ) * ( x2 - x1 )
```
[Здесь и далее] Переменные в Kotlin принято называть т. н. lower camel case'ом. Например `changeXSqr`. `change_x_sqr` или `changexsqr` - неправильный стиль.

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -93,7 +98,11 @@ fun thirdDigit(number: Int): Int = TODO()
  */
-fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
+fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int {
+    val starttime = hoursDepart * 60 + minutesDepart
```
Lower camel case

Commit:
Недоделка урок 2 задание 1

Commit:
2 урок

Commit:
Lesson 1 исправления

Commit:
Lesson 3 начало

Commit:
Lesson 2, (task1, task2)

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/lesson2/task1/IfElse.kt

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,15 +57,32 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    return when {
+        n == 1 -> 1
+        n == 2 -> 1
+        else -> fib( n - 1 ) + fib( n - 2 )
```
- Рекурсивное решение напрашивается, исходя из определения чисел Фибоначчи, но попробуйте посчитать, к скольки вызовам функции `fib()` приведёт вызов `fib(10)`, `fib(40)`, `fib(55)`, сколько среди них вызовов с одинаковыми аргументами и решить задачу более оптимальным способом.

gagarski commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +60,36 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  */
-fun angleInRadian(grad: Int, min: Int, sec: Int): Double = TODO()
-
+fun angleInRadian(grad: Int, min: Int, sec: Int): Double = grad * PI / 180 + min * PI / 10800.0 + sec * PI / 648000.0
```
Старайтесь избегать неочевидных констант, вычисленных вручную.
Человеку, читающему ваш код скорее всего будет понятно, что такое 60 и 180, а про 10800 и 648000 - уже не совсем очевидно.
Пишите прямо в коде `60 * 180` и `60 * 60 * 180`, или хотя бы поясняйте происхождение этих чисел в комментарии рядом.

gagarski commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +33,13 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
  */
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String {
+    return when {
```
Тесты для этой функции не проходят. Не все случаи рассмотрены.

mglukhikh replied:
Исправлено

mglukhikh commented:
recheck all

Commit:
Доделанный 3 урок, начало 4 урока.

gagarski commented:
recheck all

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,81 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * минимальное число k, которое делится и на m и на n без остатка
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun lcm(m: Int, n: Int): Int {
```
Почитайте про более эффективные алгоритмы вычисления НОД/НОК.

gagarski replied:
done

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -96,7 +140,14 @@ fun maxDivisor(n: Int): Int = TODO()
  * Например, 25 и 49 взаимно простые, а 6 и 8 -- нет.
  */
-fun isCoPrime(m: Int, n: Int): Boolean = TODO()
+fun isCoPrime(m: Int, n: Int): Boolean {
```
Решите эту задачу, используя НОД или НОК.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +172,22 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
  */
-fun sin(x: Double, eps: Double): Double = TODO()
+fun sin(x: Double, eps: Double): Double {
+    var currentExp = x % (2 * Math.PI) // текущее значение x^(какой-то степении) в членах ф-ции
```
Протестируйте ваши реализации синуса и косинуса на больших числах. Решите проблемы, используя свойства синуса и косинуса.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +262,23 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    var digit = 0 // номер текущей цифры в последовательности
+    var number = 0 // целое число, крадрат которого вычисляется
+    var result = 0 // n-ая цифра
+    var revertedSquare: Int // квадрат, записанный в обратном порядке
```
Зачем нужно число, записанное в обратном порядке?

gagarski replied:
актуально

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,47 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    if (v.isEmpty()) return 0.0
```
Пустой список можно не рассматривать, как особый случай. В случае пустого списка вы просто не попадёте в цикл.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,47 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isNotEmpty()) {
```
Пустой список можно не рассматривать, как особый случай. В случае пустого списка вы просто не попадёте в цикл.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +153,14 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
  */
-fun times(a: List<Double>, b: List<Double>): Double = TODO()
+fun times(a: List<Double>, b: List<Double>): Double {
+    if (a.isEmpty() && b.isEmpty()) return 0.0
```
Пустой список можно не рассматривать, как особый случай. В случае пустого списка вы просто не попадёте в цикл.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -138,7 +170,16 @@ fun times(a: List<Double>, b: List<Double>): Double = TODO()
  */
-fun polynom(p: List<Double>, x: Double): Double = TODO()
+fun polynom(p: List<Double>, x: Double): Double {
+    if (p.isEmpty()) return 0.0
```
Пустой список можно не рассматривать, как особый случай. В случае пустого списка вы просто не попадёте в цикл.

Commit:
Исправления в уроках 1-4

Commit:
Исправление задания во 2-м уроке и +2 теста к нему
Выполнение еще 2х средних задач из 4 урока

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -98,12 +99,28 @@ fun fib(n: Int): Int {
+        nn = t
     }
-    return k
+    return c / mm
```
- Лучше вынести вычисление НОД в отдельную функцию
- Лучше сделать `mm / НОД * nn` чтобы избежать переполнения например для НОК(1000000, 1000000)

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -141,12 +158,22 @@ fun maxDivisor(n: Int): Int {
+    //}
+    //return true
+    var mm = m
+    var nn = n
```
- Как сказано выше, лучше вынести НОД в отдельную функцию и использовать её здесь.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,98 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun lcm(m: Int, n: Int): Int {
+    //var k = Math.max(n, m)                   // - 1 вариант
```
Старайтесь не оставлять закомментированные части кода. Для версионирования ваших функций используйте git (который вы и так используете)

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,98 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun maxDivisor(n: Int): Int = TODO()
+fun maxDivisor(n: Int): Int {
+    var i = n - 1
```
Можно ли ограничить верхнюю границу перебора?

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,98 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun minDivisor(n: Int): Int = TODO()
+fun minDivisor(n: Int): Int {
+    var i = 2
+    while (n % i != 0) {
```
Можно ли ограничить верхнюю границу перебора (и вернуть `n`, если мы её достигли).

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -166,4 +318,20 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+    var digit = 0 // номер искомой цифры в последовательности
+    var number = 0 // целое число, fib которого вычисляется
+    var result = 0 // искомая цифра
+    var revertedFib: Int
```
Зачем нужно число, записанное в обратном порядке?

gagarski replied:
актуально

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +201,20 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+        if (nn % divider == 0) {
+            list.add(divider)
+            nn /= divider
+            divider--
```
Ваш `--` перед `++` -- пример паттерна "костыль". Почему просто не убрать инкремент в тех случаях, когда это не нужно.

gagarski replied:
(вспомните, что можно писать ветку `else` у `if`, например)

gagarski reviewed (changes requested):
Пожалуйста исправьте текущие ошибки и замечания, прежде чем приступать к следующим задачам.

Commit:
Исправления 3 урока

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,87 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * минимальное число k, которое делится и на m и на n без остатка
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun gkd(a: Int, b: Int): Int { //  доп. ф-ция для нахождения НОД, используется  в ф-ции lcm
```
Greatest Common Divisor - `gcd`

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +57,87 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun minDivisor(n: Int): Int = TODO()
+fun minDivisor(n: Int): Int {
+    for (i in 2..n / 2) { // n может нацело делиться либо на число, не больщее чем n/2, либо на самого себя (n / n = 1)
```
А можно ли ещё сильнее ограничить перебор (см. примеры к этому уроку). В `maxDivisor` тоже.

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +171,25 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var digit = 3 // для факториала
+    var xx = x
+    val absEps = Math.abs(eps) // чтобы каждый раз не вызывать в условии abs(eps)
+    if (Math.abs(x) > 2 * Math.PI) xx = x % (2 * Math.PI)
```
Нужна ли проверка условия перед взятием остатка от деления?

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +198,40 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var xx = x
+    val absEps = Math.abs(eps) // чтобы каждый раз не вызывать в пост-условии abs(eps)
+    var currentExp = 1.0 // текущее значение x^n
+    if (Math.abs(x) > 2 * Math.PI) xx = x % (2 * Math.PI)
```
Нужна ли проверка условия перед взятием остатка от деления?

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +201,20 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+        if (nn % divider == 0) {
+            list.add(divider)
+            nn /= divider
+            divider--
```
актуально

Commit:
Убрал костыль в ф-ции из 4 урока, попытался улучшить min/maxDivisor в 3ем.
Отказаться от revert в последних 2 ф-циях из 3 урока не смог, потому что не представляю, как решить иначе.
  Число записываю в обратном порядке для того, чтобы его остатком от деления на 10 стала первая цифра исходного числа. Например, из предпоследней ф-ции 4*4->16 revert -> 61 % 10 -> 1

gagarski commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +265,22 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    var revertedSquare: Long // квадрат, записанный в обратном порядке
+    while (digit < n) {
+        number++
+        revertedSquare = (revert(number * number)).toLong()
```
Поотлаживайте вашу программу на случаях, предоставленных ботом. Например, 361128-я цифра последовательности встречается в квадрате числа 40737. Его квадрат влезает в тип Int, а перевёрнутый квадрат вызывает переполнение. Попробуйте решить эту задачу без `revert`.

gagarski commented on src/lesson4/task1/List.kt:
```diff
@@ -148,7 +184,15 @@ fun polynom(p: List<Double>, x: Double): Double = TODO()
-fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+fun accumulate(list: MutableList<Double>): MutableList<Double> {
+    var sum = 0.0
+    val result = mutableListOf<Double>()
```
См. сообщение бота. Предполагается, что вы будете изменять входной список, а не создавать его изменённую копию.

VladislavMikh replied:
Возможно рассматривать в данном пулл реквесте только задачи уроков 1-3?
Мне бы хотелось закрыть его, а для проверки задач из уроков 4+ использовать другой пулл реквест.

Commit:
Полностью переделал последние 2 задачи из урока 3. Решил их без revert'а

mglukhikh commented:
recheck all

mglukhikh commented:
По просьбе автора, PR закрывается с принятием уроков 1-3. Урок 4 пока не засчитан. Просьба обратить внимание на сообщение бота для `accumulate`.
